tools/mpy_ld.py

```diff
@@ -684,6 +803,276 @@ def do_relocation_data(env, text_addr, r):
         assert 0, r_info_type
 
 
+RISCV_RELOCATIONS_TYPE_MAP = {
+    R_RISCV_ADD8: ("riscv_addsub", "B", 8, 1),
+    R_RISCV_ADD16: ("riscv_addsub", "<H", 16, 1),
+    R_RISCV_ADD32: ("riscv_addsub", "<I", 32, 1),
+    R_RISCV_ADD64: ("riscv_addsub", "<Q", 64, 1),
+    R_RISCV_SUB6: ("riscv_addsub", "B", 6, -1),
+    R_RISCV_SUB8: ("riscv_addsub", "B", 8, -1),
+    R_RISCV_SUB16: ("riscv_addsub", "<H", 16, -1),
+    R_RISCV_SUB32: ("riscv_addsub", "<I", 32, -1),
+    R_RISCV_SUB64: ("riscv_addsub", "<Q", 64, -1),
+    R_RISCV_SET6: ("riscv_set6", "B", 6),
+    R_RISCV_SET8: ("riscv_set8", "B", 8),
+    R_RISCV_SET16: ("riscv_set16", "<H", 16),
+    R_RISCV_SET32: ("riscv_set32", "<I", 32),
+    R_RISCV_JAL: "riscv_j",
+    R_RISCV_BRANCH: "riscv_b",
+    R_RISCV_RVC_BRANCH: "riscv_cb",
+    R_RISCV_RVC_JUMP: "riscv_cj",
+    R_RISCV_CALL: "riscv_call",
+    R_RISCV_CALL_PLT: "riscv_call",
+    R_RISCV_PCREL_LO12_I: "riscv_lo12i",
+    R_RISCV_PCREL_LO12_S: "riscv_lo12s",
+    R_RISCV_LO12_I: "riscv_lo12i",
+    R_RISCV_LO12_S: "riscv_lo12s",
+    R_RISCV_32_PCREL: "riscv_32pcrel",
+    R_RISCV_PLT32: "riscv_32pcrel",
+}
+
+
+def process_riscv32_relocation(env, text_addr, r):
```

Comment:
Does it make sense to split this function into "compute" and "rewrite" parts?  That might follow better the existing logic above (where this function is called) used by other archs.

Alternatively, maybe it's better to have a single function like this for all the other archs?  Although I think in their case the different archs share a lot of the rewrite logic.

Reply:
It's been a while, but the only reason was to keep the method length/complexity below the threshold set in the precommit checks running on python code.

As you point out, other archs more or less share the same relocation handling primitives when it comes to relocation handling with RISC-V being the outlier here (and MIPS and probably AArch64 too).  I'm OK with doing the split myself (either with a set of functions or classes with just enough inheritance to reuse even more code), whether that is submitted as its own PR or as an entry in this set of commits is fine - whichever way takes less effort for you to review.

Reply:
I'm happy to leave this refactoring for another day (ie keep it as you have it).

This one-file program is getting quite large so probably needs some more serious factoring, eventually.